### ISSUE TITLE
`extend range` on LFO DEFORM function

### ISSUE THREAD

**Issue by hems** [surge-synthesizer/surge#5837]
It would be great to DEFORM the LFOs a lot more than they're deformed now, if you take Serum as an example when you deform on the MOD matrix it can go to an extreme to a point where it nearly becomes a ON/OFF.

When trying to deform the LFO to use as a super "expo env" it doesn't really go there.

It would be great to be able to "right click" and do "extend range" on the DEFORM function too!

![image](https://user-images.githubusercontent.com/27327/151650220-2a71e843-9dff-4cfa-bb24-4e0af895397d.png)

Serum examples as reference:

![image](https://user-images.githubusercontent.com/27327/151650267-076ee735-2852-4f9e-b263-db442de86522.png)

![image](https://user-images.githubusercontent.com/27327/151650276-0af414d7-0ab1-4529-a2c6-d6fd30452188.png)

**Comment by hems** [comment]
I can get close with DEFORM on a triangle for instance:

![image](https://user-images.githubusercontent.com/27327/151650703-e9dd0fff-f6ef-4fd2-9103-e0fabb77dcde.png)

The problem is then that the TRIANGLE is looping. If we had a TRIGGER mode as additional mode it could be helpful..

So we would have:

 - Freerun
 - KeyTrigger
 - Random


 - **_TRIGGER_**
 - Unipolar

When TRIGGER is engaged then the LFO would not LOOP, it would only play once.

This together with "extend range" on the DEFORM would be great to make super sharp attacks and modulations similar to what i can get with DEFORM on Serum.

**Comment by hems** [comment]
Another great option could be to have that "DEFORM" on the MOD MATRIX so you could have different DEFORM per MOD, like Serum does ( :

**Comment by mkruselj** [comment]
My opinion is if you want to use Serum's features, use Serum. 🙂 

For any more intricate LFO shaping, you have the MSEG mode, where you can do things in much more detailed way than what Deform does.

**Comment by baconpaul** [comment]
deform per mod is a mod mapper by the way and we have definite plans to add those! It's a good idea.

Agree that the deform on the fixed lfo types is probably not that exciting a project considering mseg + formula give you infinite flex basically.

**Comment by hems** [comment]
> My opinion is if you want to use Serum's features, use Serum. 🙂
> 
> For any more intricate LFO shaping, you have the MSEG mode, where you can do things in much more detailed way than what Deform does. Including your suggested Trigger mode, and more exponential segments! So that about covers what you want to do.

I do use Serum a lot but I also like Surge and I feel somethings could be easily improved ( on both synths ) by "borrowing" from each other.

For instance having `extended range` on the DEFORM function ( as you do on PITCH for instance ) would be a great add considering you already have the `extend range` feature on the synth.

Closing/bombing this issue saying like "if you want to use Serum just use Serum" makes it looks like you taking it personally @mkruselj ... 

let's chill and have 🍻 and please consider 'extended range' and option to not-loop modulation waves : D

please please please : D

**Comment by baconpaul** [comment]
It’s a pretty easy request actually. Lemme reopen it and think 

but really formula and MSEG can get you there today

**Comment by hems** [comment]
> It’s a pretty easy request actually. Lemme reopen it and think
> 
> but really formula and MSEG can get you there today

the MSEG is freaking amazing indeed! i should use it more, it actually does everything i need so i understand @mkruselj closing the issue... 

In the other hand i feel like eventually every fader on Surge could have an "extended range" option when you right click it!

And having an option for "loop mode" for the other OSCs would also be very helpful for quickly baking things.

**Comment by mkruselj** [comment]
Not every parameter _needs_ extended range. We have a purppose-made tool for the things you want and it's the MSEG and it's the Formula. You can set them up however you want and save a modulator preset, then it's "quickly baked". So I really don't see much need to extend Deform...

**Comment by baconpaul** [comment]
yeah definitely we don't want extend on every param. i think that was hyperbolic not serious. "I'd like to extend the top of the cutoff to 10x nyquist" isn't *that* great a feature.

i'm sort of also of the mind that if you want this use mseg or formula.  but pondering.

**Comment by hems** [comment]
> yeah definitely we don't want extend on every param. i think that was hyperbolic not serious. "I'd like to extend the top of the cutoff to 10x nyquist" isn't _that_ great a feature.
> 
> i'm sort of also of the mind that if you want this use mseg or formula. but pondering.

agree with you and @mkruselj but even if you use MSEG the DEFORM parameter still useful and having extended range on the DEFORM in order to deform the MSEG is still a great feature IMHO.

TLDR:
 - use MSEG
 - MSEG is great
 - using deform on MSEG is great
 - Oh what if we could extend the deform so i could deform my MSEG even more
 - we back to square 0 

thanks for putting with me guys : D

🙇

**Comment by baconpaul** [comment]
yeah i think extend on deform so it goes -4 <> 4 as opposed to -1 <> 1 and make sure the LFOs don't blow up with that range is a reasonable suggestion

You'd almost definitely never use it with formula of course.

but it's easy enough to do.

(Ideally you could individually modulate points in an mseg.... but that's a can of works of epic proportion)

**Comment by baconpaul** [comment]
by the way i think deform type 2 on the envelope does what you want. Just looking at it.

**Comment by hems** [comment]
> by the way i think deform type 2 on the envelope does what you want. Just looking at it.

i tried other types and didn't get the result i was expecting.. anyway extended range should be nice for all the types?

maybe i'm bothering you guys too much... you know better than me ( :

**Comment by mkruselj** [comment]
I still stand my ground that this is not really necessary to have.

**Comment by mkruselj** [comment]
And 4 years later, after looking at the code, and attempting the modification, I think there's more complications here than it would be useful. Some modes really depend on deform staying within +/-1, others get their calculation completely thrown off instead of producing an expected result.

Closing this issue as a consequence.
